### PR TITLE
#4 - Include eslint via airbnb stlye guide

### PR DESCRIPTION
**Pull request by artpropp**
Issue #4

### CODE REVIEW

File: test/testUser.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
-'use strict';
+/* eslint-disable import/no-extraneous-dependencies */
```
Comment (DennisAhaus) [inline test/testUser.js]: Please explain why you disable this rule

File: gulpfile.js [hunk 2]
```diff
@@ -1,46 +1,46 @@
-'use strict';
-var gulp = require('gulp');
-var eslint = require('gulp-eslint');
-var excludeGitignore = require('gulp-exclude-gitignore');
-var mocha = require('gulp-mocha');
-var istanbul = require('gulp-istanbul');
-var plumber = require('gulp-plumber');
+/* eslint-disable import/no-extraneous-dependencies */
```
Comment (DennisAhaus) [inline gulpfile.js]: Why you disable this rule? This rule is very important!

File: gulpfile.js [hunk 3]
```diff
@@ -1,46 +1,46 @@
...
+  .pipe(excludeGitignore())
+  .pipe(istanbul({
+    includeUntested: true,
+  }))
+  .pipe(istanbul.hookRequire())
+));
 
-gulp.task('test', ['istanbul'], function (cb) {
-  var mochaErr;
+gulp.task('test', ['istanbul'], (cb) => {
```
Comment (DennisAhaus) [inline gulpfile.js]: `(cb) => {` --> `cb => {`

File: gulpfile.js [hunk 4]
```diff
@@ -1,46 +1,46 @@
...
+  let mochaErr;
 
   gulp.src('test/**/test*.js')
     .pipe(plumber())
     .pipe(mocha({
-      reporter: 'spec'
+      reporter: 'spec',
     }))
-    .on('error', function (err) {
+    .on('error', (err) => {
```
Comment (DennisAhaus) [inline gulpfile.js]: `(err) => {` --> `err => {`

File: lib/User.js [hunk 5]
```diff
@@ -1,26 +1,24 @@
...
-            throw new Error('Missing model parameter');
-        }
-        this._model = model;
-        this.username = null;
-        this.password = null;
+  constructor(model) {
+    if (!model) {
+      throw new Error('Missing model parameter');
     }
+    this.model = model;
```
Comment (DennisAhaus) [inline lib/User.js]: Why did you change private model property to public (without underscore)?

File: lib/index.js [hunk 6]
```diff
@@ -1,3 +1 @@
-'use strict';
```
Comment (DennisAhaus) [inline lib/index.js]: Why did you remove `use strict`? This should be in every file!

File: test/testUser.js [hunk 7]
```diff
@@ -22,23 +22,23 @@ describe(__filename, () => {
 
     it('.save() should call .save(callback) on model', () => {
       const model = {
-        save: null
+        save: null,
```
Comment (DennisAhaus) [inline test/testUser.js]: This should be an eslint error. Because in original json the last comma ist not allowed

File: test/testUser.js [hunk 8]
```diff
@@ -22,23 +22,23 @@ describe(__filename, () => {
...
       };
       const user = new User(model);
       model.save = sinon.stub();
       const returnValue = Math.random();
       model.save.yields(null, returnValue);
 
       return user.save()
-        .then(result => {
+        .then((result) => {
```
Comment (DennisAhaus) [inline test/testUser.js]: `(result) => {` --> `result => {`

File: test/testUser.js [hunk 9]
```diff
@@ -22,23 +22,23 @@ describe(__filename, () => {
...
           assert.strictEqual(result, returnValue);
           assert.equal(model.save.withArgs(sinon.match.func).callCount, 1);
         });
     });
 
     it('.save() should handle error on model error', (done) => {
       const model = {
-        save: null
+        save: null,
```
Comment (DennisAhaus) [inline test/testUser.js]: Same as above

File: test/testUser.js [hunk 10]
```diff
@@ -49,7 +49,7 @@ describe(__filename, () => {
         .then(() => {
           done(new Error('Should not reach this code'));
         })
-        .catch(err => {
+        .catch((err) => {
```
Comment (DennisAhaus) [inline test/testUser.js]: `(err) => {` --> `err => {`

File: gulpfile.js [hunk 11]
```diff
@@ -1,46 +1,46 @@
...
       mochaErr = err;
     })
     .pipe(istanbul.writeReports())
     .pipe(istanbul.enforceThresholds({
       thresholds: {
-        global: 95
-      }
+        global: 95,
```
Comment (DennisAhaus) [inline gulpfile.js]: Last comma problem

File: gulpfile.js [hunk 12]
```diff
@@ -1,46 +1,46 @@
...
-    .pipe(istanbul({
-      includeUntested: true
-    }))
-    .pipe(istanbul.hookRequire());
-});
+gulp.task('istanbul', () => (
+  gulp.src('lib/**/*.js')
+  .pipe(excludeGitignore())
+  .pipe(istanbul({
+    includeUntested: true,
```
Comment (DennisAhaus) [inline gulpfile.js]: last comma, read comments below

File: lib/User.js (hunk 5 above)
Comment (artpropp) [inline lib/User.js]: For me, the explanation in airbnb styleguide - [airbnb/javascript 22.4](https://github.com/airbnb/javascript#naming--leading-underscore) - is totally reasonable. Therefor, I implemented this rule.

File: gulpfile.js (hunk 4 above)
Comment (artpropp) [inline gulpfile.js]: I would prefer the proposed notation `(err) => {` , since it can be extended easily by further parameters, e.g. `(err, next) => {` - under the premise that the context is valid. Apart from that, it is more in line with the notation `() => {}`.
By another propose, we can follow the airbnb style guideline at [section 8.2](https://github.com/airbnb/javascript#arrows--implicit-return), and change the above mentioned block (lines 34-36) to `.on('error', err => mochaErr = err)`, which I prefer as it is a more concise approach. But again, this will leave this style question unsolved.

File: test/testUser.js (hunk 7 above)
Comment (artpropp) [inline test/testUser.js]: One argument for use of trailing commas, since they are allowed with EcmaScript5 ([quote](http://eslint.org/docs/rules/comma-dangle.html)), is to keep git diffs cleaner ([quote](https://github.com/airbnb/javascript#commas--dangling))

File: gulpfile.js (hunk 2 above)
Comment (artpropp) [inline gulpfile.js]: Without disabling this rule, not only package dependencies but also developer dependencies are allowed. Another solution would include to amend the eslint rules by `rules:
  import/no-extraneous-dependencies: [error, { devDependencies: true }]`

File: test/testUser.js (hunk 1 above)
Comment (artpropp) [inline test/testUser.js]: Without disabling this rule, not only package dependencies but also developer dependencies are allowed. Another solution would include to amend the eslint rules by `rules: import/no-extraneous-dependencies: [error, { devDependencies: true }]`

File: gulpfile.js (hunk 3 above)
Comment (artpropp) [inline gulpfile.js]: Please refer also to the comment below regarding `(err) => {`

File: test/testUser.js (hunk 10 above)
Comment (artpropp) [inline test/testUser.js]: Please refer also to the other comment below regarding `(err) => {`

File: test/testUser.js (hunk 9 above)
Comment (artpropp) [inline test/testUser.js]: Please refer to the comment regarding trailing comma above

File: gulpfile.js (hunk 12 above)
Comment (artpropp) [inline gulpfile.js]: Please refer to the comment regarding trailing comma above

File: gulpfile.js (hunk 11 above)
Comment (artpropp) [inline gulpfile.js]: Please refer to the comment regarding trailing comma above

**Comment by artpropp** [comment]
Please have also a look at the Travis-CI log output of the first commit: https://travis-ci.org/OrgApp/node-persistence-mongodb/builds/168047388

File: test/testUser.js (hunk 1 above)
Comment (DennisAhaus) [inline test/testUser.js]: Yes, it would be better to include in .eslintrc file. Do not include rules in files

File: gulpfile.js (hunk 4 above)
Comment (DennisAhaus) [inline gulpfile.js]: Okl, lets do it like the airbnb style says. This guide was made by many people and is bullet proofed i think. So please refactor as needed by the style guide. After that i can approve

File: lib/User.js (hunk 5 above)
Comment (DennisAhaus) [inline lib/User.js]: I aggree

File: lib/index.js (hunk 6 above)
Comment (DennisAhaus) [inline lib/index.js]: Ok, but what if will have 500 files in the future and decide to use another tool to compile our code BEFORE eslint runs?

File: test/testUser.js (hunk 1 above)
Comment (artpropp) [inline test/testUser.js]: Agreed

**Comment by artpropp** [comment]
What is your oppinion on trailing commas?
One argument for the use of trailing commas is to keep git diffs cleaner ([quote](https://github.com/airbnb/javascript#commas--dangling)), as trailing commas are allowed with EcmaScript5 ([quote](http://eslint.org/docs/rules/comma-dangle.html)).
But one argument against the application of trailing commas would be backwards compatibility, especially for code that should later also run on legacy browsers.

**Comment by DennisAhaus** [comment]
I think it is better not to use trailing commas because it is a cleaner style and if there is a comma you  do not know if there is a element missing (and accidently deleted) or not. Withoout that comma it is clear that this is the last element.
